src/coreclr/tools/Common/Compiler/ObjectWriter/WasmObjectWriter.cs

```diff
@@ -0,0 +1,233 @@
+
+            EmitWasmHeader(outputFileStream);
+            EmitSection(() => WriteTypeSection(uniqueSignatures, logger), outputFileStream, logger);
+            EmitSection(() => WriteFunctionSection(methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteExportSection(methodBodies.ToArray(), methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteCodeSection(methodBodies.ToArray(), logger), outputFileStream, logger);
+        }
+
+        private WasmSection WriteExportSection(IReadOnlyList<IMethodBodyNode> methodNodes, IReadOnlyList<WasmFuncType> methodSignatures,
+            Dictionary<WasmFuncType, int> signatureMap, Logger logger)
+        {
+            WasmSection exportSection = new WasmSection(WasmSectionType.Export, new SectionData(), "export");
+            SectionWriter writer = exportSection.Writer;
+            // Write the number of exports
+            writer.WriteULEB128((ulong)methodNodes.Count);
+            for (int i = 0; i < methodNodes.Count; i++)
+            {
+                var methodNode = methodNodes[i];
+                var methodSignature = methodSignatures[i];
+                string exportName = methodNode.GetMangledName(_nodeFactory.NameMangler);
+
+                var length = Encoding.UTF8.GetByteCount(exportName);
+                writer.WriteULEB128((ulong)length);
+                writer.WriteUtf8StringNoNull(exportName);
+                writer.WriteByte(0x00); // export kind: function
+                writer.WriteULEB128((ulong)i);
+                if (logger.IsVerbose)
+                {
+                    logger.LogMessage($"Emitting export: {exportName} for function index {i}");
+                }
+            }
+
+            return exportSection;
+        }
+
+        private WasmSection WriteCodeSection(IReadOnlyCollection<IMethodBodyNode> methodBodies, Logger logger)
+        {
+            WasmSection codeSection = new WasmSection(WasmSectionType.Code, new SectionData(), "code");
+
+            SectionWriter writer = codeSection.Writer;
+            // Write the number of functions
+            writer.WriteULEB128((ulong)methodBodies.Count);
+            foreach (IMethodBodyNode methodBody in methodBodies)
+            {
+                if (methodBody is ObjectNode objectNode)
+                {
+                    ObjectNode.ObjectData body = objectNode.GetData(_nodeFactory);
+
+                    writer.WriteULEB128((ulong)body.Data.Length);
+                    writer.EmitData(body.Data);
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Expected method body node to be an ObjectNode, but got {methodBody.GetType()}");
+                }
+            }
+
+            return codeSection;
+        }
+
+        private WasmSection WriteTypeSection(IEnumerable<WasmFuncType> functionSignatures, Logger logger)
```

Comment:
Here and elsewhere: there seems to be a lot of use of interfaces / abstracter collection types, but this is private code. If we have `ArrayBuilder` (or even simply `List<T>`) we can use that in these private methods and save both working set and time.